### PR TITLE
User declaration of input fields via JSDOC in script files.

### PR DESCRIPTION
**Pull request by muzam1l**
Maybe it requires a tiny spec, but it is mostly standard JSDOC https://jsdoc.app.

We only use one tag `@input` (alias `arg`, `param`, `argument`). Standard form is 
```
/**
@tag {type} name description
*/
```
Where `type` can be any html input type like `text`, `number`, etc (TODO should we disallow some?) or of the form `Account.{field}` like `Account.address`, `Account.secret`, etc. Simply having type as `Account` will return Account address. Another required field is `name`, it is the identifier for the variable and its value can be accessed in script via `process.env.{name}`, it is also the fallback for Input label which would otherwise be `description`. 

Above example constitutes a block, script can have one or multiple blocks and all of the Input tags will be merged.

Fields can be optional and have default value like.
- Optional field.
```
/**
@input {text} [gender] Type your gender.
*/
```

- Default value
```
/**
@input {text} [gender=male] Type your gender.
*/
```

While I fix some loose ends, any beta feedback welcomed too!

### CODE REVIEW

**Review by vbar — COMMENTED** [review]
I think it's fine, but since it breaks (some) hook examples, could you please update them in your gists and (temporarily) change templateFileIds to point to them before merge? That way we won't confuse beta testers and hopefully avoid extra bug reports...

**Comment by muzam1l** [comment]
Yep seems like better option. I was just waiting for official gists to be updated, guess we can point to official ones once as they are updated!

**Comment by vbar** [comment]
> Yep seems like better option. I was just waiting for official gists to be updated, guess we can point to official ones once as they are updated!

Well, the problem with updating the official gists first is that that would break the old IDE version - which would be especially inconvenient on the day the IDE is officially released... Personally, I wouldn't merge anything today - we can wait and see whether the project survives the surge of interest till Monday... :-)

**Comment by muzam1l** [comment]
Affected examples now point to forked versions, this ensures smooth transition without breaking any branch. @vbar can you test if everything is working as expected?

**Comment by vbar** [comment]
Also, in the notary example, the `{{invoice_id}}` parameter wasn't converted.

**Comment by vbar** [comment]
> > In the firewall example, I don't like that the execution mode of block-unblock.js can be any integer - it really has just 2 valid values...
> 
> That is currently a stretch, as we are making UI from user provided JSDOC!. But yeah, maybe in future we can have a Select dropdown there, but for now it works, as scripts throws on invalid exec modes with proper error message.

But users are downright encouraged to change the value outside its range... If the dropdown is too complicated, wouldn't it be better to remove the increment and decrement arrows and just have the value as a generic string (as the old version has it, too)?

File: components/RunScript/index.tsx [hunk 1]
```diff
@@ -163,17 +170,29 @@ const RunScript: React.FC<{ file: IFile }> = ({ file: { content, name } }) => {
...
     value: acc.address,
   }));
 
+  const isDisabled = Object.values(fields).some(
+    field => field.required && !field.value
+  );
+
+  const handleRun = useCallback(() => {
+    if (isDisabled)
+      return toast.error("Please fill in all the require fields.");
```
Comment (vbar) [inline components/RunScript/index.tsx:189]: I'm not a native speaker, but I think it should be "required"

File: components/RunScript/index.tsx [hunk 2]
```diff
@@ -17,16 +21,21 @@ import {
...
+const generateHtmlTemplate = (code: string, data?: Record<string, any>) => {
+  let processString: string | undefined;
+  const process = { env: { NODE_ENV: "production" } } as any;
+  if (data) {
+    Object.keys(data).forEach(key => {
+      process.env[key] = data[key];
+    });
   }
-);
+  processString = JSON.stringify(process);
```
Comment (vbar) [inline components/RunScript/index.tsx:37]: why is process stringified, only to be parsed again?

File: components/RunScript/index.tsx (hunk 1 above)
Comment (muzam1l) [reply components/RunScript/index.tsx]: Yes indeed, fixed!

File: components/RunScript/index.tsx (hunk 2 above)
Comment (muzam1l) [reply components/RunScript/index.tsx:37]: Cuz we are generating html template, which is a string, so passing data object stringified there. It's parsing is then done in script scope (when script is run by user) to be made available to user as normal object.

**Review by vbar — APPROVED** [review]

**Pull request merged**